### PR TITLE
Added AST parser

### PR DESCRIPTION
**Pull request by marcelaraujo**
Hi

I remove the current parser which uses regular expression. They don't support a few things like compressed html and other stuffs. I rewrote the `file` and `file processor` files and the tests too.

It's working as we expected.

Thanks

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
+.idea
```
Comment (sindresorhus) [inline .gitignore]: nope

File: Gruntfile.js [hunk 2]
```diff
@@ -43,5 +43,10 @@ module.exports = function (grunt) {
   grunt.loadTasks('tasks');
 
   grunt.registerTask('default', ['jshint', 'jscs', 'mochacli']);
-  grunt.registerTask('test', 'default');
+  grunt.registerTask('test', function(file) {
```
Comment (sindresorhus) [inline Gruntfile.js]: `grunt.registerTask('test', function (file) {`

File: lib/file.js [hunk 3]
```diff
@@ -33,6 +33,20 @@ var fs = require('fs');
...
+    tmp[ currentValue.name ] = currentValue.value;
+    return tmp;
+  }, tmp);
+};
+
+var camelize = function (string) {
+  return string.replace(/-+(.)?/g, function (match, chr) {
+    return chr ? chr.toUpperCase() : '';
+  });
+};
```
Comment (sindresorhus) [inline lib/file.js]: use https://github.com/sindresorhus/camelcase

File: lib/file.js [hunk 4]
```diff
@@ -33,6 +33,20 @@ var fs = require('fs');
 //    <!-- build:css /foo/css/site.css -->
 // then dest will equal foo/css/site.css (note missing trailing /)
 //
+var getAttributes = function (attributes) {
+  var tmp = {};
+  return attributes.reduce(function (previousValue, currentValue) {
+    tmp[ currentValue.name ] = currentValue.value;
```
Comment (sindresorhus) [inline lib/file.js]: `tmp[currentValue.name] = currentValue.value;`

File: .gitignore (hunk 1 above)
Comment (marcelaraujo) [inline .gitignore]: Ok.

File: Gruntfile.js (hunk 2 above)
Comment (marcelaraujo) [inline Gruntfile.js]: Keep or remove?

File: lib/file.js (hunk 3 above)
Comment (marcelaraujo) [inline lib/file.js]: Ok.

File: lib/file.js (hunk 4 above)
Comment (marcelaraujo) [inline lib/file.js]: What is it wrong?

File: lib/fileprocessor.js [hunk 5]
```diff
@@ -151,10 +151,16 @@ var FileProcessor = module.exports = function (type, patterns, finder, logcb, bl
 //
 FileProcessor.prototype.replaceBlocks = function replaceBlocks(file) {
   var result = file.content;
-  var linefeed = /\r\n/g.test(result) ? '\r\n' : '\n';
+
   file.blocks.forEach(function (block) {
-    var blockLine = block.raw.join(linefeed);
-    result = result.replace(blockLine, this.replaceWith(block));
+    var firstBlock = (block.raw[0] + '').replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
+    var lastBlock = (block.raw[block.raw.length - 1] + '').replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&');
```
Comment (sindresorhus) [inline lib/fileprocessor.js]: instead of the regexes, just use https://github.com/sindresorhus/escape-string-regexp

File: lib/file.js (hunk 4 above)
Comment (marcelaraujo) [inline lib/file.js]: What a about `data` attributes or anything else that uses hash as ng-if, my-custom-attribute?

Comment (sindresorhus) [inline lib/file.js]: I'm commenting about the code style. Nothing else.

File: Gruntfile.js (hunk 2 above)
Comment (sindresorhus) [inline Gruntfile.js]: Again. Code style.

File: lib/file.js (hunk 4 above)
Comment (marcelaraujo) [inline lib/file.js]: The jshint broken when I remove space between brackets.

File: Gruntfile.js (hunk 2 above)
Comment (marcelaraujo) [inline Gruntfile.js]: Ok.

File: lib/file.js (hunk 4 above)
Comment (marcelaraujo) [inline lib/file.js]: Done.

File: Gruntfile.js (hunk 2 above)
Comment (marcelaraujo) [inline Gruntfile.js]: Done.

**Comment by marcelaraujo** [comment]
Do these changes need a new version?

**Comment by stephanebachelier** [comment]
@marcelaraujo hmm I'm currently working too on moving from regexp to an html parser but based on htmlparser2 to fixed tons of issues on this repo... :/

Will review your PR. 
By the way you should squash your commits.

**Comment by marcelaraujo** [comment]
Ok. I will do it.

Now, about htmlparser2, I tried but it doesnt't allow you know where you need change your code. Parser5 gives to you two location variables. This is important for content replacement after all parser work.

**Comment by marcelaraujo** [comment]
I will rebase everything.

Git sometimes sucks!

**Pull request closed without merging**